### PR TITLE
HIVE-25161: Implement partitioned CTAS for iceberg tables

### PR DESCRIPTION
**Pull request by marton-bod**
- Partition columns and types should be added to the serdeProperties inside `PlanUtils.getTableDesc()`, under new `serdeConstant` keys generated by thrift
- The HiveStorageHandler had to be extended with a new method, which indicates whether to leave the partition cols in the object inspector during `FileSinkOperator.initializeOp()`. It's false by default, only overridden by Iceberg.
- In the `HiveIcebergSerDe`, we use the new partitioning info from the serdeProperties to add the partition cols to the regular table schema, as well as construct the partitioning spec.
- The `HiveStorageHandler.getLockType()` has been overridden for Iceberg so that the default exclusive lock does not clash with the table commit operation exclusive lock request; now using shared_read lock instead. Accordingly, changed the commit operation from exclusive to excl_write so that it works together with the shared_read lock.

### CODE REVIEW

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java:241]: nit: newline

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java:246]: nit: newline

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:348]: This is again something which is not by the spec. What happens if this will not propagated?

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java
Comment (pvary) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:545]: Do we want to create a method for this long line of code? We already has this 3 times

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java
Comment (pvary) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:553]: Why not use `testTables.createTable` methods?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/FileSinkOperator.java
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/FileSinkOperator.java:743]: Maybe a comment would be good here

File: ql/src/java/org/apache/hadoop/hive/ql/exec/FileSinkOperator.java
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/FileSinkOperator.java:751]: What happens if we return null here? Would the next map fail?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java
Comment (marton-bod) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java:241]: Do you mean there should be a newline added after line 241?

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java
Comment (marton-bod) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:553]: It's because I wanted to use a regular Hive table as the source here (since that should be a typical CTAS use case for Iceberg)

File: ql/src/java/org/apache/hadoop/hive/ql/exec/FileSinkOperator.java
Comment (marton-bod) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/FileSinkOperator.java:751]: If we return null in any of the map functions in Optional, then it will fall back to the `orElse` part (i.e. subsequent map functions won't fail)

File: ql/src/java/org/apache/hadoop/hive/ql/exec/FileSinkOperator.java
Comment (marton-bod) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/FileSinkOperator.java:754]: No, it shouldn't. If we return null in any of the map functions in Optional, then it will fall back to the `orElse` part (i.e. subsequent map functions won't fail)

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java
Comment (pvary) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:553]: Got it. The we might want to create a method for this as well, since we have this code at least in 3 places now

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java
Comment (marton-bod) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java:158]: That's a good question. Elsewhere in the code when parsing the list of columns from the props, they do the same thing and simply split the list by `,`  - but I guess that's no guarantee by itself. I'll try to find out what restrictions there are for column naming.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java
Comment (marton-bod) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:348]: Yes, you're right. There's actually an open upstream Iceberg PR for extracting out the serialization stuff from here: https://github.com/apache/iceberg/pull/2598. Let's wait for that to be merged then we can backport it to Hive as well.

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java
Comment (marton-bod) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:545]: Yes, that's a good idea. Will check the catalog type in the serDe, and will throw an exception with this text for non-Hive catalogs.

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java
Comment (marton-bod) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:596]: Yes, good idea. Will add one more partitioning scheme.

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java
Comment (marton-bod) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:545]: Yes, makes sense

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java
Comment (marton-bod) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:553]: Sure, makes sense

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java
Comment (pvary) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java:241]: yeah

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java
Comment (marton-bod) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:348]: You are right. However, all this code will soon be removed anyway after the table cache PR has been merged, so I would leave it as it is for now if that's OK with you.

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java
Comment (marton-bod) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:562]: That's true. However, it might not apply in this very specific case, because the target table will have a different column order than the source table (because in the `PARTITIONED BY` syntax, I flipped the column order relative to the source table), so the comparison will fail. So in this case, this check is the simplest way to do it.

Comment (marton-bod) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:562]: However, I used this method to simplify the code in two other CTAS test cases.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/FileSinkOperator.java
Comment (marton-bod) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/FileSinkOperator.java:743]: Done

Comment (marton-bod) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/FileSinkOperator.java:743]: Decided to also rename the method to `readPartitionValues()`

File: ql/src/java/org/apache/hadoop/hive/ql/plan/PlanUtils.java
Comment (marton-bod) [reply ql/src/java/org/apache/hadoop/hive/ql/plan/PlanUtils.java:267]: Good point, after some digging I've found that `MetaStoreUtils` has some util methods for this already

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java
Comment (marton-bod) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java:158]: After some digging, it seems that column names can indeed contain `,`. If they do, Hive will use a different delimiter. I've fixed the serDe code to take that into account.

However, right now Iceberg tables do not handle having column names with a comma very well (those tables can be created but can't be read back), so that is something that should be addressed in a separate PR. I've filed a jira for that.

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java
Comment (marton-bod) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:545]: Done

**Comment by marton-bod** [comment]
@lcspinter @pvary Can you please take a look again if you have any more comments?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java:128]: nit: newline after end of the block

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java
Comment (pvary) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:548]: nit: after @lcspinter's patch we can use `STORED BY ICEBERG`

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java
Comment (pvary) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:584]: nit: after @lcspinter's patch we can use STORED BY ICEBERG

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java
Comment (pvary) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:586]: Shall we use the `AssertHelpers.assertThrows`

**Review by pvary — APPROVED** [review]
small nits

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java
Comment (marton-bod) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java:128]: Done

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java
Comment (marton-bod) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:548]: Done

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java
Comment (marton-bod) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:584]: Done

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java
Comment (marton-bod) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:586]: Ah yes, that's much better

**Review by lcspinter — APPROVED** [review]

**Pull request merged**
